dot/parachain/spam_slots.go

```diff
@@ -0,0 +1,200 @@
+package parachain
+
+import (
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/parachain"
+	"github.com/emirpasic/gods/sets/treeset"
+	"sync"
+)
+
+// MaxSpamVotes is the maximum number of spam votes a validator can have for a particular session
+const MaxSpamVotes = 50
+
+// SpamSlots is an interface for managing spam votes
+type SpamSlots interface {
+	// AddUnconfirmed adds a spam vote for the given validator and candidate
+	// returns true if the vote was added, false if the validator has already voted too many times
+	// This is called for any validator's invalid vote for any not yet confirmed dispute
+	AddUnconfirmed(session parachain.SessionIndex, candidate common.Hash, validator parachain.ValidatorIndex) bool
+	// Clear out spam slots for a given candidate in a given session
+	// We call this once a dispute becomes obsolete or got confirmed and thus votes for it should no longer be treated
+	// as potential spam.
+	Clear(session parachain.SessionIndex, candidate common.Hash)
+	// PruneOld prune all spam slots for sessions older than the given index.
+	PruneOld(oldestIndex parachain.SessionIndex)
+}
+
+type slotKey struct {
+	session   parachain.SessionIndex
+	validator parachain.ValidatorIndex
+}
+
+type unconfirmedKey struct {
+	session   parachain.SessionIndex
+	candidate common.Hash
+}
+
+// spamSlots is an implementation of SpamSlots
+type spamSlots struct {
+	slots     map[slotKey]uint32
+	slotsLock sync.RWMutex
+
+	unconfirmed     map[unconfirmedKey]*treeset.Set
+	unconfirmedLock sync.RWMutex
+
+	maxSpamVotes uint32
+}
+
+// byValidatorIndex is a comparator for ValidatorIndex
+func byValidatorIndex(a, b interface{}) int {
+	return int(a.(parachain.ValidatorIndex) - b.(parachain.ValidatorIndex))
+}
+
+func slotsKey(session parachain.SessionIndex, validator parachain.ValidatorIndex) slotKey {
```

Comment:
would suggest to use `new` to create instances of a struct, same for `unconfirmedDisputesKey`
```suggestion
func newSlotKey(session parachain.SessionIndex, validator parachain.ValidatorIndex) slotKey {
```

Reply:
That's a good one, I've made the changes!